### ISSUE TITLE
Quaternions has no support for autodiff or symbolic expressions

### ISSUE THREAD

**Issue by baggepinnen** [JuliaGeometry/Quaternions.jl#8]
Definitions like `Quaternion(s::Real, v1::Real, v2::Real, v3::Real, n::Bool = false)` prevents the use of packages like `DualNumbers` to perform automatic differentiation.
http://www.juliadiff.org/ 

The type specification `::Real` further makes it impossible to use any kind of symbolic math software.

This could be worth considering to make the package useful in more situations.

**Comment by SimonDanisch** [comment]
Pull requests to change this are very welcome and easy to write :)

**Comment by bzinberg** [comment]
> Definitions like `Quaternion(s::Real, v1::Real, v2::Real, v3::Real, n::Bool = false)` prevents the use of packages like `DualNumbers` to perform automatic differentiation.

Is this actually true?  I tried to use `ForwardDiff` with `Quaternions` and only found one obstacle (granted, I didn't do thorough testing, it was a means to an end).  In particular, `ForwardDiff.Dual` _is_ a subtype of `Real`.

The issue I ran into is this line in `Quaternions.qrotation`:
https://github.com/JuliaGeometry/Quaternions.jl/blob/08d7fa7903d99c92a7e5f2a04c04a6d3011ea033/src/Quaternion.jl#L197
I was trying to differentiate with respect to `theta`, and `u` was a `Vector{Float64}`, so I got the (correct) error that `theta` could not be converted to a `Float64`.

I don't see any reason this conversion needs to happen.  Indeed, autodiff worked for me after overriding the method to not have that line.

I'll try to send in a PR.
